File: adv.py
def has_roll(grid, row, col):
    if row < 0 or col < 0 or row >= len(grid) or col >= len(grid[row]):
        return 0
    return grid[row][col] == '@'
    
def count_adj(grid, row, col, max_rolls):
    rolls = 0
    spots = [[row-1, col-1], [row-1, col], [row-1, col+1],
            [row, col-1],                   [row, col+1],
            [row+1, col-1], [row+1, col], [row+1, col+1]]
    for spot in spots:
        rolls+= has_roll(grid, spot[0], spot[1])
    return rolls < max_rolls        

allowed = 0
grid = []
with open("code.txt", "r") as f:
    for line in f:
        grid.append(list(line.strip()))

for row in range(0, len(grid)):
    for col in range(0, len(grid[row])):
        if grid[row][col] == '@':
            allowed += count_adj(grid, row, col, 4)
        

print(f"Allowed {allowed}")

File: adv.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn has_roll(grid: &Vec<Vec<char>>, row: usize, col: usize) -> isize {
    if row >= grid.len() || col >= grid[0].len(){
        return 0;
    }
    return (grid[row][col] == '@') as isize;
}

fn count_adj(grid: &Vec<Vec<char>>, row: isize, col:isize, max:isize) -> i32{
    let mut rolls = 0;
    let spots = vec![(row - 1, col - 1), (row - 1, col), (row - 1, col + 1), (row, col -1), (row, col + 1), (row + 1, col - 1), (row + 1, col), (row + 1, col + 1)];
    for (row, col) in spots{
        rolls += has_roll(grid, row as usize, col as usize);
    }
    return (rolls < max) as i32;
}

fn main() -> io::Result<()> {
    
    let reader = io::BufReader::new(File::open(Path::new("code.txt"))?);
    let mut grid: Vec<Vec<char>> = Vec::new();
    for line in reader.lines() {
        let line = line?;
        let row: Vec<char> = line.chars().collect();
        grid.push(row);
    }
    let mut allowed = 0;
    for row in 0..grid.len(){
        for col in 0..grid[row].len(){
            if grid[row][col] == '@'{
                allowed += count_adj(&grid, row as isize, col as isize, 4);
            }
        }
    }
    println!("Allowed {}", allowed);
    return Ok(());
}
